Fix some edge cases and add unit tests for `checkErrors` function

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

For the `checkErrors` functions at `.~/components/free-listings/configure-product-listings/checkErrors.js`

- Add unit tests
- Fix some edge cases to make sure the merchant settings data called to API will not contain invalid values

### Screenshots

https://user-images.githubusercontent.com/17420811/125068092-a548a200-e0e7-11eb-9619-c541a328387f.mp4

### Detailed test instructions:

1. Run `npm run test-unit` and make sure all these tests pass
2. Delete or move out the `gla_merchant_center` WP options to load the initial MC onboarding setup from scratch
3. Proceed to step 3
    - The "Complete setup" button should be disabled if the free shipping threshold is not yet entered
    - Other checks should work as before

### Changelog entry

> Fix - Disable the "Complete setup" button if the free shipping price is not yet entered when setting up Merchant Center for the first time

eason9487 committed:
Add test cases for checkErrors function

eason9487 committed:
Check invalid values in checkErrors function

tomalec commented on js/src/components/free-listings/configure-product-listings/checkErrors.test.js:
```diff
@@ -0,0 +1,352 @@
+			manualShipping = { shipping_rate: 'manual' };
+		} );
+
+		it( 'When the type of shipping rate is an invalid value, should not pass', () => {
```
:nail_care: 
WDYT of adding "or missing" to explicitly match what's tested here.
```suggestion
		it( 'When the type of shipping rate is an invalid value or missing, should not pass', () => {
```

Also, we could describe, what "not pass" means. Like, 
```suggestion
		it( 'When the type of shipping rate is an invalid value or missing, should not pass - returned object should contain `shipping_rate` property with an error message.', () => {
```
We could add that to JSDoc as well.

eason9487 replied:
The descriptions of "returned object should contain property" are applicable for all individual checks, so I added a comment note on the top level of the test scope. Adjusted by https://github.com/woocommerce/google-listings-and-ads/pull/871/commits/5101eb382c404e929cb0efaab698b064f8af418e

tomalec commented on js/src/components/free-listings/configure-product-listings/checkErrors.test.js:
```diff
@@ -0,0 +1,352 @@
+				};
+			}
+
+			it( `When there are any selected countries' shipping rates is not set, should not pass`, () => {
```
:nail_care: 
I'm not an English expert, but, I'm not sure if that's grammatically correct.

```suggestion
			it( `When there are any selected countries with shipping rates not set, should not pass`, () => {
```

tomalec commented on js/src/components/free-listings/configure-product-listings/checkErrors.test.js:
```diff
@@ -0,0 +1,352 @@
+
+				const errors = checkErrors( flatShipping, rates, [], codes );
+
+				expect( errors ).not.toHaveProperty( 'shipping_rate' );
```
:nail_care: 
I'm not sure how it affects tests atomicity:
With the way, I understand what "pass" means for `checkErrors` in general - there is no property with an error message.
Then we should either assert for `errors` not to have **any** property or change the description to specify there should be no rate error `shipping_rate`.

I could imagine :somebo: could expect that in some case `errors` could have another property here like `errors.missing_country`.

eason9487 replied:
Adjusted comment by https://github.com/woocommerce/google-listings-and-ads/pull/871/commits/5101eb382c404e929cb0efaab698b064f8af418e and added test case by https://github.com/woocommerce/google-listings-and-ads/pull/871/commits/2e4494a9fa0309998e4640aa48ea8dc91c618d9e
(:somebo: 😂 )

tomalec commented on js/src/components/free-listings/configure-product-listings/checkErrors.test.js:
```diff
@@ -0,0 +1,352 @@
+				expect( errors.shipping_rate ).toMatchSnapshot();
+			} );
+
+			it( `When all shipping rates are ≥ 0, should pass`, () => {
```
> ≥ 

:sunglasses:

tomalec commented on js/src/components/free-listings/configure-product-listings/checkErrors.test.js:
```diff
@@ -0,0 +1,352 @@
+			} );
+
+			it( 'When the free shipping threshold is an invalid value, should not pass', () => {
+				// Not set yet
+				let freeShipping = createFreeShipping( null );
```
:nail_care: 
For not set we could add also `undefined`

eason9487 replied:
Added a comment about the `null` value by https://github.com/woocommerce/google-listings-and-ads/pull/871/commits/a66e87b7f0db763b5de329109e31f0b93f26fb05

tomalec reviewed (approved):
Besides some cosmetic and editorial comments, LGTM

eason9487 committed:
English grammatical correction

Co-authored-by: Tomek Wytrębowicz <tomalecpub@gmail.com>

eason9487 committed:
Adjust test descriptions to clarify the test intentions

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/871#discussion_r670659892
- https://github.com/woocommerce/google-listings-and-ads/pull/871#discussion_r670666081

eason9487 committed:
Assert checkErrors for the returned value when checks are all passed or partial unpassed

Address https://github.com/woocommerce/google-listings-and-ads/pull/871#discussion_r670666081

eason9487 committed:
Add comment to clarify the not set yet value of free_shipping_threshold is null

Address https://github.com/woocommerce/google-listings-and-ads/pull/871#discussion_r670667868
